sdk/samples/iot/sample.c

```diff
@@ -0,0 +1,454 @@
+  }
+
+  // Free the BIO chain
+  BIO_free_all(source_mem_bio);
+
+  return result;
+}
+
+az_result hmac_sha256_sign(const az_span* decoded_key, const az_span* signature, az_span* out_span)
+{
+  PRECONDITION_NOT_NULL(decoded_key);
+  PRECONDITION_NOT_NULL(signature);
+  PRECONDITION_NOT_NULL(out_span);
+
+  az_result result;
+
+  unsigned int hmac_encode_len;
+  unsigned char* hmac = HMAC(
+      EVP_sha256(),
+      (void*)az_span_ptr(*decoded_key),
+      az_span_size(*decoded_key),
+      az_span_ptr(*signature),
+      (size_t)az_span_size(*signature),
+      az_span_ptr(*out_span),
+      &hmac_encode_len);
+
+  if (hmac != NULL)
+  {
+    *out_span = az_span_create(az_span_ptr(*out_span), (int32_t)hmac_encode_len);
+    result = AZ_OK;
+  }
+  else
+  {
+    result = AZ_ERROR_INSUFFICIENT_SPAN_SIZE;
+  }
+
+  return result;
+}
+
+az_result base64_encode(const az_span* bytes, az_span* out_span)
+{
+  PRECONDITION_NOT_NULL(bytes);
+  PRECONDITION_NOT_NULL(out_span);
+
+  az_result result;
+  BIO* sink_mem_bio;
+  BIO* b64_encoder;
+  BUF_MEM* encoded_mem_ptr;
+
+  // Create a BIO filter to process the bytes
+  b64_encoder = BIO_new(BIO_f_base64());
+  if (b64_encoder == NULL)
+  {
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+
+  // Create a memory sink BIO to process bytes to
+  sink_mem_bio = BIO_new(BIO_s_mem());
+  if (sink_mem_bio == NULL)
+  {
+    BIO_free(b64_encoder);
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+
+  // Push the sink to the encoder
+  b64_encoder = BIO_push(b64_encoder, sink_mem_bio);
+  if (b64_encoder == NULL)
+  {
+    BIO_free(sink_mem_bio);
+    BIO_free(b64_encoder);
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+
+  // Set no newline flag for the encoder
+  BIO_set_flags(b64_encoder, BIO_FLAGS_BASE64_NO_NL);
+
+  // Write the bytes to be encoded
+  int bytes_written = BIO_write(b64_encoder, az_span_ptr(*bytes), (int)az_span_size(*bytes));
+  if (bytes_written < 1)
+  {
+    BIO_free(sink_mem_bio);
+    BIO_free(b64_encoder);
+    return AZ_ERROR_OUT_OF_MEMORY;
+  }
+
+  // Flush the BIO
+  BIO_flush(b64_encoder);
+
+  // Get the pointer to the encoded bytes
+  BIO_get_mem_ptr(b64_encoder, &encoded_mem_ptr);
+
+  if ((size_t)az_span_size(*out_span) >= encoded_mem_ptr->length)
+  {
+    // Copy the bytes to the output and initialize output span
+    memcpy(az_span_ptr(*out_span), encoded_mem_ptr->data, encoded_mem_ptr->length);
+    *out_span = az_span_create(az_span_ptr(*out_span), (int32_t)encoded_mem_ptr->length);
+
+    result = AZ_OK;
+  }
+  else
+  {
+    result = AZ_ERROR_INSUFFICIENT_SPAN_SIZE;
+  }
+
+  // Free the BIO chain
+  BIO_free_all(b64_encoder);
+
+  return result;
+}
+
+void sas_generate_encoded_signed_signature(
+    const az_span* sas_key,
+    const az_span* sas_signature,
+    az_span* sas_b64_encoded_hmac256_signed_signature)
+{
+  PRECONDITION_NOT_NULL(sas_key);
+  PRECONDITION_NOT_NULL(sas_signature);
+  PRECONDITION_NOT_NULL(sas_b64_encoded_hmac256_signed_signature);
+
+  int rc;

```

Comment:
why is this 'rc' here and 'result' in other places?

Reply:
thanks for the catch.  Will search and replace with rc.